src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1259,39 +1262,53 @@ def rename_domain(self, old, new):
         Dict(
             'vm_delete',
             Bool('zvols', default=False),
+            Bool('force', default=False),
         ),
     )
-    def do_delete(self, id, data):
+    async def do_delete(self, id, data):
         """Delete a VM."""
-        self.ensure_libvirt_connection()
-        status = self.status(id)
-        if status.get('state') == 'RUNNING':
-            self.poweroff(id)
-        elif status.get('state') == 'ERROR':
-            raise CallError('Unable to retrieve VM status. Failed to destroy VM')
-
-        if data['zvols']:
-            devices = self.middleware.call_sync('vm.device.query', [
-                ('vm', '=', id), ('dtype', '=', 'DISK')
-            ])
+        async with LIBVIRT_LOCK:
+            vm = await self._get_instance(id)
+            self.ensure_libvirt_connection()
```

Comment:
Is it safe to run ensure_libvirt_connection synchronously in the event loop?

Reply:
I think so but maybe i can still separate this to be certain we don't get into any issues.

Reply:
Actually no, we are checking connection.isalive - definitely updating this